certbot-nginx/certbot_nginx/parser.py

```diff
@@ -325,19 +345,22 @@ def create_new_vhost_from_default(self, vhost_template):
         """
         # TODO: https://github.com/certbot/certbot/issues/5185
         # put it in the same file as the template, at the same level
+        all_vhosts = self.get_vhosts()
         enclosing_block = self.parsed[vhost_template.filep]
         for index in vhost_template.path[:-1]:
             enclosing_block = enclosing_block[index]
-        new_location = vhost_template.path[-1] + 1
         raw_in_parsed = copy.deepcopy(enclosing_block[vhost_template.path[-1]])
-        enclosing_block.insert(new_location, raw_in_parsed)
+        enclosing_block.append(raw_in_parsed)
         new_vhost = copy.deepcopy(vhost_template)
-        new_vhost.path[-1] = new_location
-        for addr in new_vhost.addrs:
-            addr.default = False
-        for directive in enclosing_block[new_vhost.path[-1]][1]:
-            if len(directive) > 0 and directive[0] == 'listen' and 'default_server' in directive:
-                del directive[directive.index('default_server')]
+        new_vhost.path[-1] = len(enclosing_block) - 1
+        if delete_default:
+            for addr in new_vhost.addrs:
+                addr.default = False
+            for directive in enclosing_block[new_vhost.path[-1]][1]:
+                if len(directive) > 0 and directive[0] == 'listen' \
+                    and 'default_server' in directive:
+                    del directive[directive.index('default_server')]
+        all_vhosts.append(new_vhost)
```

Comment:
The cache is actually not for performance reasons, it's to be able to start keeping around within-run information (namely `ssl_copy`). For a while, it's been bothering me that we re-parse and find server blocks anew every time, even when we want to have found the same server block, instead of just keeping a reference to it. But we couldn't do that before because `get_vhosts()` was ephemeral. I think that this is a cleaner, safer way to be doing it generally.

If we didn't cache them, we'd certainly have to change the `use_if` as you mentioned above.

But, that being said, the way for users of this class to create vhosts is to call this method, which handles updating the cache. They'd also have to update the raw blocks, which they also wouldn't know to do. I think adding a comment saying "this method is the only way to create a new vhost" would suffice.

Reply:
> But, that being said, the way for users of this class to create vhosts is to call this method, which handles updating the cache. They'd also have to update the raw blocks, which they also wouldn't know to do. I think adding a comment saying "this method is the only way to create a new vhost" would suffice.

When I initially reviewed this PR I think I accidentally misread the code and thought `duplicate_vhost` was a standalone function and updating this cache was done outside of the parser itself. At the very least, that's what I remembered until I looked at the code again. Technically that wouldn't matter, but stylistically doing that with no documentation made my soul hurt. Luckily, I read it wrong.

This is fine, but I'd err on the side of too much documentation here as I think this has a chance to cause bugs in the future as this code is no longer in recent memory. Places to think about putting documentation are the docstrings of the parser class, `get_vhosts`, and `duplicate_vhost`. Inline comments about this might not hurt either.